### PR TITLE
Integrate scheduler into the compiler

### PR DESCRIPTION
**Pull request by BoxiLi**
**Description**
- Integrate scheduler into the compiler. The compiler can now schedule quantum gates to reduce the duration of the compiled pulses.
- Improve the structure and readability of the `Compiler` class. The compiler class was separated for clarity when the `Processor` class was built upon old code in `qutip.qip`. But it has never been opened as a public API. This PR improves the readability and, most importantly, makes it easier for users to customize the compiler class.
- Add a map `pulse_dict` between the pulse label and the indices in `Processor`, so that customizing class becomes much easier.

**Files description**
- `scheduler.py` is moved to compiler folder and split to two files : `instruction.py` and `scheduler.py` for clarity.
- Changes in `gatecompiler.py`, `cavityqedcompiler.py` and `spinchaincompiler.py` are for integrating the scheduler.
- Others are adaptations

**Changelog**
Integrate the scheduler into the compiler

### CODE REVIEW

**Comment by BoxiLi** [comment]
The failing test has nothing to do with this PR

File: qutip/tests/test_compiler.py
Comment (jakelishman) [inline qutip/tests/test_compiler.py:40]: you're importing `QubitCircuit` twice?

File: qutip/tests/test_compiler.py
Comment (jakelishman) [inline qutip/tests/test_compiler.py:87]: seems odd to split the imports up like this

File: qutip/tests/test_compiler.py
Comment (jakelishman) [inline qutip/tests/test_compiler.py:99]: continuous (extra u) (and kwargs don't usually have space around the equals)

**Review by jakelishman — CHANGES_REQUESTED** [review]
Honestly, I'm not qualified to properly review this.  I spotted several things that "smell" wrong to me, but I'm really not familiar with the `qip` module, so I basically can't comment on architecture.

**Comment by BoxiLi** [comment]
> I spotted several things that "smell" wrong to me, but I'm really not familiar with the qip module, so I basically can't comment on architecture.

Thanks for the effort! Anything hints on what could be wrong? I can give them more thought. No need to be very specific or after 
 careful thinking, feelings also wellcome.

**Comment by jakelishman** [comment]
@BoxiLi: I meant the things I commented on inline in the code

**Comment by BoxiLi** [comment]
Oh I see, thanks, I'm going through them.

**Comment by jakelishman** [comment]
don't feel like I'm definitely correct in my comments - I'm not familiar with the code, so if there's a reason that I'm wrong just say.

File: qutip/qip/compiler/gatecompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:54]: I checked the [numpy doc style again](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy-style-python-docstrings), commonly used docstyle is:
```
    **kwargs:
        Key word arguments for ...
```
It's ok to not have a type but I'm missing here the `**`. Added

File: qutip/qip/compiler/gatecompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:44]: Yes, you are right.  ``:class:`.QubitCircuit' `` ``:class:`.Gate' `` work well.

File: qutip/qip/compiler/gatecompiler.py
Comment (jakelishman) [reply qutip/qip/compiler/gatecompiler.py:54]: that would be when documentating a constructor, right?  The constructor here doesn't take `**kwargs`, and if it's intended to be an attribute on `self`, it'd need the type (`**kwargs` in a call unambiguously binds a `dict` to `kwargs`, but an attribute `self.kwargs` could be anything).

File: qutip/qip/compiler/cavityqedcompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/cavityqedcompiler.py:149]: This was a mistake, if it is a rectangular pulse, it can be array-like or just a number. Probably forget to delete `0.,`

File: qutip/qip/device/processor.py
Comment (BoxiLi) [reply qutip/qip/device/processor.py:270]: This is `tol` is for `tlist` rather than for `Qobj` It is used to remove float-point inaccuracy when defining and concatenating `tlist`.

Comment (jakelishman) [reply qutip/qip/device/processor.py:270]: sure, my question was more whether you'd want to have `tol=None` in the call, and set `tol = tol or processor_options.get('tol')` in the function, so the defaults can be set in a centralised manner?  I don't really know if it's appropriate in this use-case, just I saw a hard-coded default.

File: qutip/qip/pulse.py
Comment (BoxiLi) [reply qutip/qip/pulse.py:421]: I forget why I put True here... The purpose is to make sure that it is larger than the `tol` so that the first element is always kept. I changed it to `10 * tol` now.

Comment (jakelishman) [reply qutip/qip/pulse.py:421]: Yeah, much better to have an explicit number that's larger than `tol`.  In theory, using `True` won't always succeed - if `tol` is larger than 1, for example.  The call still feels rather strange to me - the signature is meant to be `np.append(array, values)`, so it seems almost a bug that the call works with the array being a scalar! - but that may be preferable to the alternative (avoiding the sentinel value)
```python
full_tlist = np.unique(...)
full_tlist = np.concatenate((full_tlist[:1], full_tlist[1:][np.diff(full_tlist) > tol]))
```

File: qutip/qip/compiler/gatecompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:153]: Yes there is a mistake here.

`num_pulses`  is not the number of instructions but the number of available control Hamiltonians. It counts the used control pulses index and gets the total number. I changed the name to `num_controls` to make this clear in the new changes, as well as tests for this particular case.

The corrected code
```
num_controls = 0
for instruction in instruction_list:
    for pulse_index, _ in instruction.pulse_info:
        num_controls = max(num_controls, pulse_index)
num_controls += 1
```

File: qutip/qip/compiler/gatecompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:220]: When initializing `compiled_tlist` I made sure that it is never empty: `
compiled_tlist = [[[0.]] for tmp in range(num_controls)]
`

The 3-level nested list is for 
- control_pulse_index
- instruction
- tlist for each instruction


Here I need to know where the last time step ends. And I don't want to check every time if this is empty. So I added a 0 there and if it is no instruction for that control pulse, I'll get a 0.

File: qutip/qip/compiler/gatecompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:258]: I don't have a strong argument. I guess I made it like this because, in the end, they will be given to `QobjEvo`, which prefer `None` than `[]` for an empty `tlist`.

File: qutip/qip/device/processor.py
Comment (BoxiLi) [reply qutip/qip/device/processor.py:270]: Ah, I see. In indeed I agree that making it adjustable somewhere will be good. There are a few of this `tol` in `pulse.py` as well. The reason I'm being lazy about this is that many of them is a temporary solution for QuTiP 4. I remember with `Coefficient` in `dev.major`, `QobjEvo` with different `tlist` can be added. So they will no longer be needed.

File: qutip/qip/compiler/gatecompiler.py
Comment (jakelishman) [reply qutip/qip/compiler/gatecompiler.py:217]: It's not super obvious that the initial value you assign to the output variable `compiled_tlist` should be important, since the rest of the loop is about overwriting all of that, and three-level nested loops are difficult enough at the best of times.

For a maintainer looking at the loop in six months' time it seems like a classic off-by-one access error on the first step, especially as the initialisation is about 30--40 lines higher up.  It'd be clearer if the last time step was stored in a variable that's just called `last_step_end` or something which is initialised to `0` before the loop, not least because then you can read words rather than array accesses and you won't need a special sentinel value in the initialisation.

I get that it works now, but it just seems kind of fragile and difficult to read the next time someone needs to modify the code.

**Comment by BoxiLi** [comment]
@jakelishman Thanks for the effort you put in reviewing this! I have made the suggested changes.

There might be minor conflicts in merging these PRs as they all modify `processor.py`. I can resolve them myself once one of the PRs is approved and merged.

File: qutip/qip/device/processor.py
Comment (jakelishman) [reply qutip/qip/device/processor.py:267]: Yeah, effectively the time-dependence handling is moved out of `QobjEvo` and into `Coefficient` which makes things easier to combine/reuse things

File: qutip/qip/compiler/gatecompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:217]: > It'd be clearer if the last time step was stored in a variable that's just called last_step_end or something which is initialised to 0 before the loop

This sounds prominent. I'll work on this.

File: qutip/qip/pulse.py
Comment (BoxiLi) [reply qutip/qip/pulse.py]: Yeah, `np.append` here does seems to be a strange usage. I guess NumPy automatically promotes the first value to an array. To prevent future NumPy updates from breaking this, yours is safer.

Update made.

**Comment by BoxiLi** [comment]
Hi, @jakelishman do you have further comments on this PR or you also find this one ok to go? I can also ask others to check this if you think you are not familiar with this or it may take too much time :)

**Review by nathanshammah — APPROVED** [review]
Thanks @BoxiLi  for this major improvement. It LGTM. I suggest adding information in the documentation about this, especially on stuff like MyCompiler.

**Comment by BoxiLi** [comment]
Thanks! Documentation are added in this PR https://github.com/qutip/qutip-doc/pull/127

File: qutip/qip/device/processor.py
Comment (jakelishman) [inline qutip/qip/device/processor.py:743]: Why does the solver variable get overridden with a function value in one case but not the other?

File: qutip/qip/pulse.py
Comment (jakelishman) [inline qutip/qip/pulse.py:182]: Something's very strange in the implementation if a public attribute is a private class like this.

File: qutip/tests/test_compiler.py
Comment (jakelishman) [inline qutip/tests/test_compiler.py:102]: don't you want to test ALAP too?

File: qutip/qip/compiler/cavityqedcompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/cavityqedcompiler.py:101]: They are defined as part of hardware modelling in `Processor` (method `set_up_params`) and used here. But indeed it is probably better to make it clear in the doc of every particular subclass which parameters are required.

File: qutip/qip/compiler/gatecompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:181]: Indeed this is the difference.

I'm not sure if this is a good design so I'm just writing things down to clarify:

The intention behind `pulse_dict` is to have a map between the string name of a pulse and its location in the list `Processor.pulses`, so that while defining gate compilation, one can use string names instead of numbers. More intuitive. In all predefined `Processor`s, this map is defined in the `Processor` subclass and passed to the compiler when calling it. If `pulse_dict` is not given, it should use the integer index.

That said, I don't see it particular useful here to partially overriding this. Instead, if a `pulse_name` is not found in the dictionary, it is more likely a result of a typo: the name is mistyped. Allow it to go further will probably make it harder to debug: a string is assigned to something called `pulse_ind`.

Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:181]: Unifying the type of `pulse_dict` to `dict` does sounds nice. Maybe I can add a type check for `pulse_ind` and get rid of `pulse_dict=None`

File: qutip/qip/pulse.py
Comment (BoxiLi) [reply qutip/qip/pulse.py:182]: That's a fair point. I plan to replace it with `QobjEvo` at some point in a different PR. Make it private is a warning for that change... The difficulty is that `Pulse` does not have the `dim` info for the total Hamiltonian system. It just save the local Hamiltonian and the target qubits.

File: qutip/qip/compiler/gatecompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:195]: Sounds reasonable. Do you think they should be private class methods or separated private functions? They don't use private at all. But if put outside the function, they may end up very far away from where they used if new class methods are added.

File: qutip/qip/compiler/gatecompiler.py
Comment (jakelishman) [reply qutip/qip/compiler/gatecompiler.py:181]: Sure, if it's got to be either all of them or none of them then this makes sense.  A comment up top wouldn't go amiss - because of Python's weak "truthiness", you'd usually expect `{}` and `None` to behave the same if they're both valid inputs.

I'm not wild about the idea of exposing a public attribute that has variable type (`dict` vs `None`), but it's a minor thing and it's definitely up to you.

File: qutip/qip/compiler/gatecompiler.py
Comment (jakelishman) [reply qutip/qip/compiler/gatecompiler.py:195]: To some degree, the idea of separating them out is _so_ the "detail" code doesn't distract you from the high-level overview.  Any decent IDE can jump to the definition if you click on the name, so once it's beyond "not literally in the same place", I think it's fine as long as it's in the same file.

My tendency is always to ask "why use a class method when a function would do?", but it's really more about how you want things.  A couple of minor arguments in favour of separate functions: it compels you to think about what data really needs to be passed between each step (you can't just share a huge state via `self`), which forces individual complexity down; and there's one less level of indentation, so everything's just that little bit easier to read.  In favour of class methods: as you say, it keeps the code a little more local; it doesn't "leak" helper functions into a larger namespace (no matter that they're not exported further).

File: qutip/qip/pulse.py
Comment (jakelishman) [reply qutip/qip/pulse.py:182]: It shows a problem with some part of the process - either here or in `QobjEvo`.  In a stricter language, you wouldn't even be able to access the type `_EvoElement` in this module, because it's not part of the public API.

I don't know what exactly to suggest (I'm fairly sure this PR isn't the one that introduced its usage), but using a class marked private is rather fragile.  The next person to edit `QobjEvo` may look at `_EvoElement` and assumed (somewhat correctly!) that they're completely free to change it, because it's not part of any API boundary.

File: qutip/qip/compiler/gatecompiler.py
Comment (BoxiLi) [reply qutip/qip/compiler/gatecompiler.py:181]: After a second thought on this. I changed it and use `pulse_dict={}` to replace `pulse_dict=None`. In the end, an empty `pulse_dict` doesn't mean anything useful here. It is going to always through an error. Using it instead of `pulse_dict=None` may save some handling with `None` as you suggested in a few comments above.

File: qutip/qip/device/processor.py
Comment (BoxiLi) [reply qutip/qip/device/processor.py:322]: Good catch. Indeed the logic is similar to `pulse.py`, I think I'll leave this for this PR because (1) In QuTiP 5.0 they will be greatly simplified when the solver allows different `tlist` for different `coeff`; (2）It can be done together with the leakage of private class in `pulse.py`.

File: qutip/qip/compiler/instruction.py
Comment (BoxiLi) [reply qutip/qip/compiler/instruction.py:109]: That's a piece of very useful information. Thanks! 
I realized that this part was something left from a previous implementation and is not used any more so I removed it.

**Comment by BoxiLi** [comment]
@jakelishman Shall I rebase it after the review or better merge the master into this branch now to resolve the conflicts? The conflicts are minor and I'm pretty clear what caused them.

**Comment by jakelishman** [comment]
I guess you need to fix the merge conflict before the tests can run.  I won't have time to look through this today, so there's plenty of time.

**Review by jakelishman — APPROVED** [review]

**Comment by jakelishman** [comment]
I'll have to trust you mostly on actual operation of the code, but it looks like you've covered everything I could see as particular problems.  I don't use this part of QuTiP, so I'm really not the best placed person to talk about how natural the APIs seem or anything like that.

I'm still a fan of fixing the issues CodeClimate is complaining about because I think the cognitive complexity tests are a semi-decent indicator of maintainability, but at the end of the day, you're the code owner and that's your call.  The other seemingly major potential headache is the dependence on that internal `_EvoElement` API in `QobjEvo` - I know for sure that that's going to go to hell in 5.0 - but I get that this PR didn't really cause this, so that's maybe a problem for another time.

**Comment by BoxiLi** [comment]
Thanks, @jakelishman. Even you say that you didn't look into very details, your comments were very inspiring and helpful!

What code climate is complaining about is mostly left from the previous code structure. I do plan on further cleaning them. 

Actually, the `_EvoElement` here is a simplified "copy" of the `EvoElement` in `QobjEvo`, not directly referring to that. Switch to `dev.major` won't break it. I did this long ago because I was having problems initialize `QobjEvo` in my particular use case. But indeed exposing a private class is fragile and dangerous. Planning on remove it too.

**Comment by jakelishman** [comment]
Oh, that's good news about `_EvoElement`.  Yeah, go ahead and merge it in when you're happy with it.

**Pull request merged**
